### PR TITLE
Feature/setup queue for asynchronous tasks

### PR DESCRIPTION
**Pull request by maismin**
The current implementation uses one message queue for web scraping and an event listener that runs after a product's details is scraped from its link.

~~The current implementation uses two message queues:~~

~~1.  For web scraping product links~~
~~2. For updating products in database~~

TO DO:

- [x] Replace the message queue for updating products with an event listener.

- [x] Setup repeating jobs instead of adding them back to the queue after they are completed.

- [x] Find the best way to integrate the message queue into the server:
> No, the client shouldn't wait for the job to be completed. All the server should do is enqueue the job and return confirmation that it was enqueued. The result will be transmitted via socket.io
> Both the message queue and event listener can be outside of the controller. The controller imports the queue to add a job. The event listener imports a socket.io service that notifies the user

* Should the message queue be imported into the product's controller file?
  * ~~If so, should a callback (with the req, res, and next arguments) be attached to the data object in the queue.add() method? That way, the response will be sent from the event listener when the new product is added in the database.~~
  * ~~If not, given that an event listener was used how should one go about this problem?~~
* ~~Should the message queue be inside the product's controller file?~~
  * ~~If so, the event listener also contains logic for creating notifications and requesting users to fetch the new notifications from the server. This seems outside the scope of the product controller file.~~

### CODE REVIEW

**Comment by aeciorc** [comment]
> should a callback (with the req, res, and next arguments) ?

No, the client shouldn't wait for the job to be completed. All the server should do is enqueue the job and return confirmation that it was enqueued. The result will be transmitted via socket.io

Both the message queue and event listener can be outside of the controller. The controller imports the queue to add a job. The event listener imports a socket.io service that notifies the user

File: server/app/services/web-scraper/wss.js [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        0,
+        51,
+      )}\n`,
+    );
+    // TEST - increased currentPrice by $10 so a later fetch will update the price in DB
+    productDetails.currentPrice += 10.0;
+    let product = await Product.create(productDetails);
+    product = product.toJSON();
+
+    await webScrapeQueue.add(
```
Comment (aeciorc) [inline server/app/services/web-scraper/wss.js:85]: Does this promise resolve after the job is added or after the job is complete?

Comment (maismin) [reply server/app/services/web-scraper/wss.js:85]: I tested with the following code
```
const Queue = require('bull');

const myQueue = new Queue('my-queue');

myQueue.process(async job => {
  console.log(`RUNNING JOB with message: ${job.data.msg}`);
});

myQueue.on('completed', (job, result) => {
  console.log('Job Completed');
});

(async () => {
  console.log('Adding job to queue');

  await myQueue.add({ msg: 'TASK1' });

  console.log('Job added to queue');
})();
```
The promise resolves after the job is added. As it currently is written, the ``await`` could be left out. However, the service  would crash if the server attempted to add a message to the queue and Redis is not running. Perhaps I should wrap the 
``await myQueue.add({ msg: 'TASK1' });``
inside a try/catch block. Would that be needed for this project?

Comment (aeciorc) [reply server/app/services/web-scraper/wss.js:85]: I think that's a good idea

File: server/app/_data/users.json [hunk 2]
```diff
@@ -0,0 +1,32 @@
+[
```
Comment (Esodey) [inline server/app/_data/users.json:1]: This will be great going forward

File: server/app/models/Product.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+      type: String,
+      required: [true, 'Please add the product name'],
+    },
+    link: {
+      type: String,
+      required: [true, 'Please add the product link'],
+    },
+    imageUrl: {
+      type: String,
+      default: '',
```
Comment (Esodey) [inline server/app/models/Product.js:21]: @aeciorc brought this up with me. Do we want to have a default image until the user sets one? Would we default it here or on the frontend?

**Review by Esodey — APPROVED** [review]

File: server/app/models/Product.js (hunk 3 above)
Comment (maismin) [reply server/app/models/Product.js:21]: We can have the frontend show a default image if the string is empty. The default image can reside inside the /public/assets directory.

**Pull request merged**